### DISCUSSION TITLE
Are there sensors in Prefect?

### DISCUSSION THREAD

**Discussion by promsoft** [PrefectHQ/prefect#3249] [Q&A] [answered]
I can't find the equivalent for Airflow Sensors in Prefect documentation. Is it a feature available?

Sensors are a special kind of operator. When they run, they will check to see if a certain criteria is met before they complete and let their downstream tasks execute. This is a great way to have portions of your DAG wait on some external system.

**Comment by cicdw** [comment, marked as answer]
Hi @promsoft - because Prefect does not require a well-defined schedule for Flow Runs + Flow Runs can be parametrized, we haven't (yet) had the need to implement such a concept.  For event-driven work, we currently recommend parametrizing your Flow by event and using the API to kick off ad-hoc runs with the appropriate event payload as the events occur.  

For a concrete example of what I mean, check out this post on triggering Flow runs when objects are written to an S3 bucket: https://medium.com/the-prefect-blog/event-driven-workflows-with-aws-lambda-2ef9d8cc8f1a

**Reply by waffletower-reify** [reply]
The solution presented above works nicely when a single external event (in the example -- when data is written to an S3 bucket) is utilized to trigger a Prefect Flow.   Yes, you have idempotency keys, but I don't see how they address race conditions without providing explicit logic in the lambda code.  This isn't a big issue.

However, what of the use case where there are multiple independent event sources required to complete before a Prefect flow is triggered?  For example, a use case where multiple reports from different vendors (perhaps in entirely different buckets and accounts) need to arrive before a flow is executed.   It seems that this use case is not supported without the dependency logic being handled external to Prefect.  Airflow's Sensor pattern handles this use case via *pull polling*  (I would personally prefer *push notification* support in an orchestration environment, particularly for S3 Event Notifications)
